### ISSUE TITLE
Add double-click hinting mode

### ISSUE THREAD

**Issue by The-Compiler** [The-Compiler/qutebrowser#446]
```
16:32 <mkonig> Is there some kind of double click?
16:33 <The-Compiler> mkonig: I don't get the question, sorry :)
16:34 <mlisin> I think he wants double-click without mouse
16:37 <mkonig> mlisin, The-Compiler : mlisin is right. I tried to open a picture on dropbox webpage. If you go to the album view than doubleblicking opens a picture. Single only un-/selects it.
```

**Comment by peterlvilim** [comment]
Did some research on this.  The problem is actually that there is an underlying hover for the area that is to be clicked.  The logic in hints.py:385 has the hover take precedence so the click isn't being sent through.  This would make sense for normal websites where you would like to hover the menu.  But fails in this case.

Also, the ondblclick isn't being detected, but in this case that isn't necessary.  Probably should be supported.

So I would propose.  Detect single click cases, double click cases, and hover cases.  Then for your hints use a different color for the hint style to indicate to the user what the hint will do if they execute it.  Use can disable hint types or change colors in settings.  Let me know your thoughts.

Above solution would resolve one of the annoyances I often run into with other types of keyboard driven browsers.

**Comment by The-Compiler** [comment]
> Did some research on this. The problem is actually that there is an underlying hover for the area that is to be clicked. The logic in hints.py:385 has the hover take precedence so the click isn't being sent through. This would make sense for normal websites where you would like to hover the menu. But fails in this case.
> 
> Also, the ondblclick isn't being detected, but in this case that isn't necessary. Probably should be supported.

I don't follow, and I believe you're misinterpreting how the current hint implementation works ;)

The current hint targets (members of the `Target` enum) are not selected automatically in any way, they are selected *by the user*, i.e. the user decides if they want to click/doubleclick/hover/download/... an element.

For example, `f` is click, `;h` is hovering, and `;D` could be double-clicking.

This brings be back to the idea of detecting those cases automatically, which I didn't even think of yet!

> So I would propose. Detect single click cases, double click cases, and hover cases. Then for your hints use a different color for the hint style to indicate to the user what the hint will do if they execute it (could put the color codes in the status bar to remind the user while hinting). The user can disable hint types or change colors in settings. Let me know your thoughts.
> 
> Above solution would resolve one of the annoyances I often run into with other types of keyboard driven browsers. I also don't think it would increase clutter because I believe currently that these areas are getting multiple hint tags (see the menus at the top of imdb.com for instance). However, I'm guessing on the clutter comment and haven't verified it unlike the rest that I wrote above.

I like this idea, but I'm not sure if it can be realized. We'd probably need `QWebElement::evaluateJavascript` to check for `on*` handlers, if that's possible at all (I suck at Javascript). There I'm not 100% certain if that'll work if javascript is disabled for a page. I believe I've seen it working before, but other sources say it shouldn't.

**Comment by peterlvilim** [comment]
That is an excellent feature that I missed!

I'll look into it.

However, based on looking at the code I still believe there is a bug in the current form.  Because if there is an area that is hoverable and clickable and if you use f on it then you'll end up hovering instead of clicking.  This is due to line hints.py:385 and I removed it to confirm this was a problem on dropbox.  Do you agree?

If so I'll submit a pull request to fix this, but I figured I would check first since I missed the hovering key bind so maybe I'm missing something else.

**Comment by The-Compiler** [comment]
I'm afraid I still don't follow. Could you elaborate on the issue you're seeing on Dropbox?

Let me explain what happens if the user presses `f`:

* `:hint` without arguments is invoked
* This means `hints.py:start` will be called with the `target` parameter set to the default, `Target.normal`.
* This will set `self._context.target` to `Target.normal`. `self._context` is later passed as `context` parameter to `_click()`.
* `_click()` will *always* click, as that's what the user said to do. Target is not set from the webelement in any way (we don't check if something is hoverable).

When the user presses `;h`, `:hint all hover` is executed instead, so the target is set to `Target.hover`, so `_click()` will get that target, and hover.

**Comment by The-Compiler** [comment]
As a note to myself, this is the patch I've had lying around which doesn't work properly:

```diff
diff --git a/qutebrowser/browser/hints.py b/qutebrowser/browser/hints.py
index 6e2d551..1a75285 100644
--- a/qutebrowser/browser/hints.py
+++ b/qutebrowser/browser/hints.py
@@ -24,7 +24,8 @@ import functools
 import subprocess
 import collections
 
-from PyQt5.QtCore import pyqtSignal, pyqtSlot, QObject, QEvent, Qt, QUrl
+from PyQt5.QtCore import (pyqtSignal, pyqtSlot, QObject, QEvent, Qt, QUrl,
+                          QTimer)
 from PyQt5.QtGui import QMouseEvent, QClipboard
 from PyQt5.QtWidgets import QApplication
 from PyQt5.QtWebKit import QWebElement
@@ -41,8 +42,8 @@ ElemTuple = collections.namedtuple('ElemTuple', ['elem', 'label'])
 
 Target = usertypes.enum('Target', ['normal', 'tab', 'tab_bg', 'window', 'yank',
                                    'yank_primary', 'run', 'fill', 'hover',
-                                   'rapid', 'rapid_win', 'download',
-                                   'userscript', 'spawn'])
+                                   'doubleclick', 'rapid', 'rapid_win',
+                                   'download', 'userscript', 'spawn'])
 
 
 @pyqtSlot(usertypes.KeyMode)
@@ -121,6 +122,7 @@ class HintManager(QObject):
         Target.run: "Run a command on a hint...",
         Target.fill: "Set hint in commandline...",
         Target.hover: "Hover over a hint...",
+        Target.doubleclick: "Doubleclick a hint...",
         Target.rapid: "Follow hint (rapid mode)...",
         Target.rapid_win: "Follow hint in new window (rapid mode)...",
         Target.download: "Download hint...",
@@ -374,14 +376,22 @@ class HintManager(QObject):
             QMouseEvent(QEvent.MouseMove, pos, Qt.NoButton, Qt.NoButton,
                         Qt.NoModifier),
         ]
-        if target != Target.hover:
+        if target not in (Target.hover, Target.doubleclick):
             self.set_open_target.emit(target.name)
+        if target != Target.hover:
             events += [
                 QMouseEvent(QEvent.MouseButtonPress, pos, Qt.LeftButton,
                             Qt.NoButton, Qt.NoModifier),
                 QMouseEvent(QEvent.MouseButtonRelease, pos, Qt.LeftButton,
                             Qt.NoButton, Qt.NoModifier),
             ]
+
+        self._emit_mouse_events(events)
+        #if target == Target.doubleclick:
+        #    QTimer.singleShot(10, functools.partial(
+        #        self._emit_mouse_events, events))
+
+    def _emit_mouse_events(self, events):
         for evt in events:
             self.mouse_event.emit(evt)
 
@@ -648,6 +658,7 @@ class HintManager(QObject):
                 - `tab-bg`: Open the link in a new background tab.
                 - `window`: Open the link in a new window.
                 - `hover` : Hover over the link.
+                - `doubleclick`: Doube-click the element.
                 - `yank`: Yank the link to the clipboard.
                 - `yank-primary`: Yank the link to the primary selection.
                 - `run`: Run the argument as command.
@@ -773,6 +784,7 @@ class HintManager(QObject):
             Target.rapid: self._click,
             Target.rapid_win: self._click,
             Target.hover: self._click,
+            Target.doubleclick: self._click,
             # _download needs a QWebElement to get the frame.
             Target.download: self._download,
             Target.userscript: self._call_userscript,
```

**Comment by rcorre** [comment]
You need this to open files/folders on google drive. It complains about using an 'unsupported browser', but things seem alright aside from not being able to open files without using the mouse (I even tried double-hinting with `--rapid` :grin:)

**Comment by uhs-robert** [comment]
I know that this is an old thread and I apologize in advance if this is not the correct forum for this comment. I recently ran into this issue with Google Drive where items need to be doubleclicked to open them. Research lead me here. 

I created a workaround keybind setup which allows double-clicking items, perhaps this may be useful:

```python
#### Double Click Mode
RESET_HINT_FOLLOW = f"set hints.auto_follow unique-match" # Replace unique-match with your preferred default
ENABLE_DBL_CLICK = "set hints.auto_follow never"
config.bind(";D;", f"{ENABLE_DBL_CLICK} ;; hint --rapid focusbox tab-bg")
config.bind(";Df", f"{ENABLE_DBL_CLICK} ;; hint --rapid all tab-bg")
config.bind("<Return>", "cmd-repeat 2 hint-follow", mode="hint")  # Press enter after selecting a hint to double-click it
config.bind(
    "<Ctrl+Return>",
    f"cmd-repeat 2 hint-follow ;; mode-leave ;; {RESET_HINT_FOLLOW}",
    mode="hint",
)
config.bind("<Escape>", f"mode-leave ;; {RESET_HINT_FOLLOW}", mode="hint")
```

Since Google Drive was also mentioned in this thread, I am using a custom hint group to target the Drive items:
```
### - Focusable containers -
c.hints.selectors["focusbox"] = [
    "main",
    "article",
    "section",
    '[role="main"]',
    '[role="article"]',
    '[role="region"]',
    "[tabindex]",
    '[contenteditable="true"]',
]
```
